spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/test/java/org/springframework/cloud/gcp/sql/autoconfig/GcpCloudSqlAutoConfigurationTests.java

```diff
@@ -0,0 +1,158 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.sql.autoconfig;
+
+import java.io.IOException;
+
+import javax.sql.DataSource;
+
+import com.google.api.services.sqladmin.SQLAdmin;
+import com.google.api.services.sqladmin.model.DatabaseInstance;
+import com.zaxxer.hikari.HikariDataSource;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.cloud.gcp.core.autoconfig.GcpContextAutoConfiguration;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author João André Martins
+ */
+@RunWith(SpringRunner.class)
+@SpringBootTest(
+		classes = {GcpCloudSqlAutoConfiguration.class, GcpContextAutoConfiguration.class},
+		properties = {"spring.cloud.gcp.projectId=proj",
+				"spring.cloud.gcp.sql.instanceName=test-instance",
+				"spring.cloud.gcp.sql.databaseName=test-database",
+				"spring.cloud.gcp.sql.userName=watchmaker",
+				"spring.cloud.gcp.sql.password=pass"}
+)
+public class GcpCloudSqlAutoConfigurationTests {
+
+	@Autowired
+	private DataSource dataSource;
+
+	@MockBean
+	private SQLAdmin mockSqlAdmin;
+
+	@Mock
+	private SQLAdmin.Instances mockInstances;
+
+	@Mock
+	private SQLAdmin.Instances.Get mockGet;
+
+	private DatabaseInstance databaseInstance = new DatabaseInstance();
+
+	@Before
+	public void setUp() throws IOException {
```

Comment:
@artembilan I'm having some issues here and could use your help.

The `cloudSqlDataSource()` bean in `GcpCloudSqlAutoConfiguration` is being configured before the `@Before` block is run, therefore, `sqlAdmin.instances()` blows up with an NPE.

I'm not sure what the best way to proceed here, since I seem to be following everything from https://spring.io/blog/2016/04/15/testing-improvements-in-spring-boot-1-4 (except I'm not testing as a web application but rather just pulling the bean from the boot app, but it should be the same).

Also, it seems that the only way to test with different property settings is to have multiple test classes for multiple property test cases? The only place I've seen where I can specify properties is at the class level...